cpp/src/arrow/util/compression_zlib.cc

```diff
@@ -47,6 +47,238 @@ static constexpr int GZIP_CODEC = 16;
+                  uint8_t* output, int64_t* bytes_read, int64_t* bytes_written) override;
+
+  Status Flush(int64_t output_len, uint8_t* output, int64_t* bytes_written,
+               bool* should_retry) override;
+
+  Status End(int64_t output_len, uint8_t* output, int64_t* bytes_written,
+             bool* should_retry) override;
+
+ protected:
+  Status ZlibError(const char* prefix_msg) {
+    std::stringstream ss;
+    ss << prefix_msg;
+    if (stream_.msg && *stream_.msg) {
+      ss << stream_.msg;
+    } else {
+      ss << "(unknown error)";
+    }
+    return Status::IOError(ss.str());
+  }
+
+  z_stream stream_;
+  bool initialized_;
+};
+
+Status GZipCompressor::Compress(int64_t input_len, const uint8_t* input,
+                                int64_t output_len, uint8_t* output, int64_t* bytes_read,
+                                int64_t* bytes_written) {
+  DCHECK(initialized_) << "Called on non-initialized stream";
+
+  stream_.next_in = const_cast<Bytef*>(reinterpret_cast<const Bytef*>(input));
+  stream_.avail_in = static_cast<uInt>(input_len);
+  stream_.next_out = reinterpret_cast<Bytef*>(output);
+  stream_.avail_out = static_cast<uInt>(output_len);
+
+  int64_t ret = 0;
+  ret = deflate(&stream_, Z_NO_FLUSH);
+  if (ret == Z_STREAM_ERROR) {
+    return ZlibError("zlib compress failed: ");
+  }
+  if (ret == Z_OK) {
+    // Some progress has been made
+    *bytes_read = input_len - stream_.avail_in;
+    *bytes_written = output_len - stream_.avail_out;
+  } else {
+    // No progress was possible
+    DCHECK_EQ(ret, Z_BUF_ERROR);
+    *bytes_read = 0;
+    *bytes_written = 0;
+  }
+  return Status::OK();
+}
+
+Status GZipCompressor::Flush(int64_t output_len, uint8_t* output, int64_t* bytes_written,
+                             bool* should_retry) {
+  DCHECK(initialized_) << "Called on non-initialized stream";
+
+  stream_.avail_in = 0;
+  stream_.next_out = reinterpret_cast<Bytef*>(output);
+  stream_.avail_out = static_cast<uInt>(output_len);
+
+  int64_t ret = 0;
+  ret = deflate(&stream_, Z_SYNC_FLUSH);
+  if (ret == Z_STREAM_ERROR) {
+    return ZlibError("zlib flush failed: ");
+  }
+  if (ret == Z_OK) {
+    *bytes_written = output_len - stream_.avail_out;
+  } else {
+    DCHECK_EQ(ret, Z_BUF_ERROR);
+    *bytes_written = 0;
+  }
+  // "If deflate returns with avail_out == 0, this function must be called
+  //  again with the same value of the flush parameter and more output space
+  //  (updated avail_out), until the flush is complete (deflate returns
+  //  with non-zero avail_out)."
+  *should_retry = (*bytes_written == 0);
+  return Status::OK();
+}
+
+Status GZipCompressor::End(int64_t output_len, uint8_t* output, int64_t* bytes_written,
+                           bool* should_retry) {
+  DCHECK(initialized_) << "Called on non-initialized stream";
+
+  stream_.avail_in = 0;
+  stream_.next_out = reinterpret_cast<Bytef*>(output);
+  stream_.avail_out = static_cast<uInt>(output_len);
+
+  int64_t ret = 0;
```

Comment:
Why on 2 lines?

Reply:
Nothing special, it's the style used (arbitrarily) throughout this PR.